### PR TITLE
feat(cli): Extract - Flatten ICU messages

### LINKED ISSUE #1186: Suggestion: option to flatten ICU Messages
FormatJS's `extract` command has an option called `--flatten`. The documentation is self-explanatory so I'll paste it here:

> Whether to hoist selectors & flatten sentences as much as possible. E.g:
> `"I have {count, plural, one{a dog} other{many dogs}}"`
> becomes `"{count, plural, one{I have a dog} other{I have many dogs}}"`.
> The goal is to provide as many full sentences as possible since fragmented
> sentences are not translator-friendly.

Do you think it would make sense to have something similar in Lingui?

### PR DESCRIPTION
**Pull request by Martin005**
# Description
Fixes #1186 

Based on suggestion from #1186, this PR adds flattening of ICU messages for the `extract` and `extract-template` CLI commands.

## Types of changes

[//]: # (What types of changes does your code introduce to Lingui?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [X] I have read the [CONTRIBUTING](https://github.com/lingui/js-lingui/blob/main/CONTRIBUTING.md) and [CODE_OF_CONDUCT](https://github.com/lingui/js-lingui/blob/main/CODE_OF_CONDUCT.md) docs
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by andrii-bodnar — APPROVED** [review]

**Review by thekip — CHANGES_REQUESTED** [review]
How this supposed to work in general? When the message would be flattened, it would be a different string. And it would not match to id string from runtime.

File: packages/core/src/flatten/flattenMessage.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        const newValue = flattenTokens([
+          ...tokens.slice(0, i),
+          ...cases[index].tokens,
+          ...tokens.slice(i + 1),
+        ])
+        all[index] = { ...k, tokens: newValue }
+        return all
+      }, [])
+
+      return [cloned]
```
Comment (thekip) [inline packages/core/src/flatten/flattenMessage.ts:32]: Without `reduce`, mutating data and `cloneDeep`: 

```ts
return [{
  ...token,
  cases: token.cases.map((k, index) => {
    const newValue = flattenTokens([
      ...tokens.slice(0, i),
      ...cases[index].tokens,
      ...tokens.slice(i + 1)
    ])

    return {...k, tokens: newValue}
  })
}]
```

**Comment by thekip** [comment]
@Martin005 

> How this supposed to work in general? When the message would be flattened, it would be a different string. And it would not match to id string from runtime.

I see here a couple ways how this could be done: 

1. This logic should be applied in the macro and not in extractor. This way macro will generate flatten messages, left them in the bundle where extractor would pick them up. But we should consider SWC version of macro, and porting these changes to the SWC would be hard (because it depends on `@message-parser`). So i would not recommend this option. 

2. Implement it together with hash based id proposal. This way a hash would be generated by macro from the original, untouched message. And extractor with `--flat` would process only "messages" not ids. The only one problem here - this will not work for users who not used macro and not use explicit ids.  Actually, for v4 we can force explicit ids to be used for non-macro usage.

**Comment by Martin005** [comment]
@thekip I have been thinking and I believe that the second option would be the best way to go.

**Comment by thekip** [comment]
Yes, this is exactly how i'm see implementation of hash-based ids with macro. 

Two cases: 

1. Without macro, `i18n._()` call, id is required. Message would not be used as substitution.
2. Macro `t` calls will generate hash based ids from message if explicit ids is not specified, and this behavior would be enabled as default.

**Pull request closed without merging**